plugins/isecl/tools/kbstool.go

```diff
@@ -0,0 +1,387 @@
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/intel-secl/intel-secl/v4/pkg/lib/common/setup"
+	"github.com/intel-secl/intel-secl/v4/pkg/model/kbs"
+)
+
+const KBS_URL = "https://localhost:9443/kbs/v1"
+const CMS_URL = "https://localhost:8445/cms/v1"
+
+func main() {
+	var keyTransferPolicy string
+	var keyID string
+	var cmd string
+	var caDir string
+	var caCertFile string
+	var tacCertFile string
+	var tacKeyFile string
+	var label string
+	flag.StringVar(&keyTransferPolicy, "ktp-id", "", "Key transfer Policy ID")
+	flag.StringVar(&keyID, "key-id", "2561fb12-c3dc-4ca8-a7e6-66abd515f7cf", "Key ID")
+	flag.StringVar(&label, "label", "test key", "Key label to use for newly created key")
+	flag.StringVar(&cmd, "cmd", "search", "possible commands: create, search, download-cms-ca-cert, download-cert, transfer and hsm-transfer")
+	flag.StringVar(&caDir, "ca-dir", "/etc/tac/", "location of the ca certificate to store/load")
+	flag.StringVar(&tacCertFile, "cert-file", "/etc/tac/client.pem", "location of the client certificate to store/load")
+	flag.StringVar(&tacKeyFile, "key-file", "/etc/tac/key.pem", "location of the client private key to load for signing")
+	flag.StringVar(&caCertFile, "ca-cert-file", "/etc/tac/ca.pem", "location of the CA certificate")
+
+	flag.Parse()
+
+	token := os.Getenv("BEARER_TOKEN")
+	if token == "" {
+		fmt.Println("ERR: Missing KBS token. Provide it via BARER_TOKEN environment!")
+		os.Exit(1)
+	}
+
+	var err error
+	if cmd == "create" {
+		err = createKey(token, keyTransferPolicy, label)
+	} else if cmd == "delete" {
+		err = deleteKey(token, keyID)
+	} else if cmd == "search" {
+		err = searchKeys(token)
+	} else if cmd == "transfer" {
+		//id := "c7470638-be5b-44de-8972-ff8f26d89880"
+		err = transferKey(token, keyID)
+	} else if cmd == "hsm-transfer" {
+		err = hsmTransfer(token, keyID, tacCertFile, tacKeyFile, caCertFile, "")
+	} else if cmd == "download-cms-ca-cert" {
+		hash := os.Getenv("CMS_CERT_HASH")
+		if hash == "" {
+			fmt.Println("ERR: Missing CMS CA certificate hash. Provide it via CMS_CERT_HASH environment. You can get it by running 'cms tlscertsha384'.")
+			os.Exit(1)
+		}
+		err = downloadCaCert(hash, caDir)
+	} else if cmd == "download-cert" {
+		hash := os.Getenv("CMS_CERT_HASH")
+		if hash == "" {
+			fmt.Println("ERR: Missing CMS CA certificate hash. Provide it via CMS_CERT_HASH environment. You can get it by running 'cms tlscertsha384'.")
+			os.Exit(1)
+		}
+		err = downloadCert(token, caDir, tacKeyFile, tacCertFile)
+	} else {
+		err = fmt.Errorf("Unrecognized command")
+	}
+	if err != nil {
+		fmt.Printf("ERR: Failed to execute command '%v': %v", cmd, err)
+	}
+}
+
+func hsmTransfer(token, keyID, certFile, keyFile, caCertFile, sessionID string) error {
+	url := KBS_URL + "/keys/" + keyID + "/dhsm2-transfer"
```

Comment:
This is bit of a magic function... you could add comments about what you are actually fetching and where does this URL come from.

Reply:
Added comments to functions in this file. Hope this helps to get an overview of the code.